go/pkg/casng/streaming_upload.go

```diff
@@ -478,8 +479,133 @@ func (u *uploader) callBatchUpload(ctx context.Context, bundle uploadRequestBund
 // For files above the large threshold, this call assumes the io and large io holds are already acquired and will release them accordingly.
 // For other files, only an io hold is acquired and released in this call.
 func (u *uploader) streamer() {
+	log.V(1).Info("[casng] upload.streamer.start")
+	defer log.V(1).Info("[casng] upload.streamer.stop")
+
+	// Unify duplicate requests.
+	digestTags := make(map[digest.Digest][]string)
+	digestReqs := make(map[digest.Digest][]string)
+	streamResCh := make(chan UploadResponse)
+	pending := 0
+	for {
+		select {
+		// The dispatcher closes this channel when it's done dispatching, which happens after the streamer
+		// had sent all pending responses.
+		case req, ok := <-u.streamerCh:
+			if !ok {
+				return
+			}
+			isLargeFile := req.reader != nil
+			log.V(3).Infof("[casng] upload.streamer.req; digest=%s, large=%t, req=%s, tag=%s, pending=%d", req.Digest, isLargeFile, req.id, req.tag, pending)
+
+			reqs := digestReqs[req.Digest]
+			reqs = append(reqs, req.id)
+			digestReqs[req.Digest] = reqs
+			tags := digestTags[req.Digest]
+			tags = append(tags, req.tag)
+			digestTags[req.Digest] = tags
+			if len(tags) > 1 {
+				// Already in-flight. Release duplicate resources if it's a large file.
+				log.V(3).Infof("[casng] upload.streamer.unified; digest=%s, req=%s, tag=%s, bundle=%d", req.Digest, req.id, req.tag, len(tags))
+				if isLargeFile {
+					u.ioThrottler.release()
+					u.ioLargeThrottler.release()
+				}
+				continue
+			}
+
+			var name string
+			if req.Digest.Size >= u.ioCfg.CompressionSizeThreshold {
+				log.V(3).Infof("[casng] upload.streamer.compress; digest=%s, req=%s, tag=%s", req.Digest, req.id, req.tag)
+				name = MakeCompressedWriteResourceName(u.instanceName, req.Digest.Hash, req.Digest.Size)
+			} else {
+				name = MakeWriteResourceName(u.instanceName, req.Digest.Hash, req.Digest.Size)
+			}
+
+			pending += 1
+			// Block the streamer if the gRPC call is being throttled.
+			startTime := time.Now()
+			if !u.streamThrottle.acquire(u.ctx) {
```

Comment:
For the context, it is more convenient to use it that way and it saves me from adding another field to track a "done" uploader. I'll address this anyway as I prefer to adhere to the style guide.

Reply:
For the throttler, I found the weighted semaphore library too heavy for what I need. The much simpler custom implementation also allows me to track the count of acquired tokens which is not straightforward to implemented for an external library.

Reply:
I wonder whether the perf difference (weighted semaphore vs custom throttler) is measurable. The extra code comes with more maintenance burden, but if we get a measurable perf boost from using the lighter version, then it might be justified.